File: leet-code/973. K Closest Points to Origin/KClosestPointsToOrigin973_2.java
package com.example;

import java.util.Comparator;
import java.util.PriorityQueue;

public class KClosestPointsToOrigin973_2 {

    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<int[]> pq = new PriorityQueue<>(Comparator.comparingDouble(this::getDistance));
        for (int[] point : points) {
            pq.offer(point);
        }
        int[][] result = new int[k][2];
        for (int i = 0; i < k; i++) {
            result[i] = pq.poll();
        }
        return result;
    }

    private double getDistance(int[] point) {
        return Math.sqrt(Math.pow(point[0], 2) + Math.pow(point[1], 2));
    }
}

File: leet-code/973. K Closest Points to Origin/KClosestPointsToOrigin973_2.kt
package com.example

import java.util.PriorityQueue

class KClosestPointsToOrigin973_2 {
    fun kClosest(points: Array<IntArray>, k: Int): Array<IntArray> {
        val pq = PriorityQueue<Node>(Comparator.comparingInt { it.distance })
        points.forEach { pq.add(Node(it)) }
        val answer = mutableListOf<IntArray>()
        IntArray(k).forEach { _ -> answer.add(pq.poll().point) }
        return answer.toTypedArray()
    }

    class Node(val point: IntArray) {
        val distance: Int = point[0] * point[0] + point[1] * point[1]
    }
}
